packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -252,40 +275,36 @@ export class TokenDetectionController extends PollingControllerV1<
   async detectTokens(options?: {
     networkClientId?: NetworkClientId;
     accountAddress?: string;
-  }) {
-    const { networkClientId, accountAddress } = options || {};
-    const {
-      disabled,
-      isDetectionEnabledForNetwork,
-      isDetectionEnabledFromPreferences,
-    } = this.config;
+  }): Promise<void> {
+    const { networkClientId, accountAddress } = options ?? {};
     if (
-      disabled ||
-      !isDetectionEnabledForNetwork ||
-      !isDetectionEnabledFromPreferences
+      this.#disabled ||
+      !this.#isDetectionEnabledForNetwork ||
+      !this.#isDetectionEnabledFromPreferences
     ) {
       return;
     }
-    const { tokens } = this.getTokensState();
-    const selectedAddress = accountAddress || this.config.selectedAddress;
-    const chainId = this.getCorrectChainId(networkClientId);
-
-    const tokensAddresses = tokens.map(
+    const { tokens } = this.#getTokensState() ?? {};
+    const selectedAddress = accountAddress ?? this.#selectedAddress;
+    const chainId = this.#getCorrectChainId(networkClientId);
+    const tokensAddresses = tokens?.map(
       /* istanbul ignore next*/ (token) => token.address.toLowerCase(),
     );
-    const { tokenList } = this.getTokenListState();
+    const { tokenList } = await this.messagingSystem.call(
+      'TokenListController:getState',
+    );
     const tokensToDetect: string[] = [];
     for (const address of Object.keys(tokenList)) {
       if (!tokensAddresses.includes(address)) {
         tokensToDetect.push(address);
       }
     }
     const sliceOfTokensToDetect = [];
-    sliceOfTokensToDetect[0] = tokensToDetect.slice(0, 1000);
-    sliceOfTokensToDetect[1] = tokensToDetect.slice(
-      1000,
-      tokensToDetect.length - 1,
-    );
+    for (let i = 0; i < tokensToDetect.length; i += 1000) {
+      sliceOfTokensToDetect.push(
+        tokensToDetect.slice(i, Math.min(i + 1000, tokensToDetect.length)),
+      );
+    }
```

Comment:
Hmm, I'm having to parse this a bit, and I feel like the original code was easier to understand (although of course it can be slightly refactored):

``` suggestion
    const sliceOfTokensToDetect = [
      tokensToDetect.slice(0, 1000),
      tokensToDetect.slice(1000)
    ];

Reply:
The new code enables `detectTokens()` to process any number of tokens as per our discussion during standup. 

Using the original code would be a regression to the previous behavior of only allowing the method to process 2 batches of 1000 tokens. If we want to preserve that restriction because it's safer or more efficient, we can definitely restore the original code.

Reply:
Ah... that's true. I guess this is a long-standing bug.

Is it necessary to make this fix in order to convert this controller to BaseControllerV2? I'm curious whether we should do this in a separate PR. I suspect there is not a proper test for this.

Reply:
This can be done in a separate PR. I'll restore the original code and create a ticket for this change + writing tests.

Reply:
- Reverted: https://github.com/MetaMask/core/pull/3609/commits/b9cbcba9f6877f4158623487f91a82d1b90ab24a
- Ticket created: https://github.com/MetaMask/core/issues/3661